### PR TITLE
Rename cEntity:m_IsSubmerged to m_IsUnderWater

### LINKED ISSUE #3987: Rename cEntity::m_IsSubmerged to m_IsHeadInWater
The member variable should be renamed to match its intended purpose. The getter should be renamed, too, but needs to keep the old Lua binding as well as the new, with the old one emitting a log message with a lua stack trace (see src/Bindings/ManualBindings.cpp)

### PR DESCRIPTION
**Pull request by bearbin**
And related changes, including manual bindings changes. Fixes #3987 

This has been tested in game and appears to work.

Please don't merge this, I will rebase and merge this manually once it's approved.

### CODE REVIEW

**Comment by bearbin** [comment]
I have updated this PR with new code, and would like a review.

Particularly, does it matter to modify the state of the entity during the checks before finally setting the variables to their final value at the end, as I am doing here? If these functions are only called during the tick thread, I can't see any issue but if it's better I can change them.

Also, Entities/Pickup.cpp line 159 checks for fire/lava both within the block of the pickup and below in order to delete it. Could this be replaced with a simple check of IsInLava() || IsInFire()?

File: Server/Plugins/APIDump/APIDesc.lua [hunk 1]
```diff
@@ -3886,6 +3926,16 @@ local Hash = cCryptoHash.sha1HexString("DataToHash")
 					},
 					Notes = "Returns true if the entity represents a {{cTNTEntity|TNT entity}}",
 				},
+				IsUnderWater =
```
[the remark this answers is not in the document]
Comment (bearbin) [reply Server/Plugins/APIDump/APIDesc.lua]: Nope, that's a remnant of the old code, I'll remove it.

File: src/Entities/Entity.cpp [hunk 2]
```diff
@@ -1647,37 +1602,70 @@ bool cEntity::MoveToWorld(const AString & a_WorldName, bool a_ShouldSendRespawn)
...
+		{
+			for (int y = MinY; y <= MaxY; y++)
+			{
+				BLOCKTYPE Block;
+				if (!a_Chunk.UnboundedRelGetBlockType(x, y, z, Block))
+				{
+					LOGD("SetSwimState failure: RelX = %d, RelY = %d, RelZ = %d, Pos = %.02f, %.02f}",
+						x, y, z, GetPosX(), GetPosZ()
+					);
+					return;
```
[the remark this answers is not in the document]
Comment (bearbin) [reply src/Entities/Entity.cpp]: Yes, that should be fine, I'll still log the error anyway.

File: src/Bindings/ManualBindings.cpp [hunk 3]
```diff
@@ -3956,6 +3956,56 @@ static int tolua_cCompositeChat_UnderlineUrls(lua_State * tolua_S)
...
+	}
+
+
+	// API function deprecated:
+	LOGWARNING("cEntity:IsSubmerged() is deprecated. Use cEntity:IsHeadInWater() instead.");
+	cLuaState::LogStackTrace(tolua_S);
+
+	L.Push(Mtolua_new((bool)(self->IsHeadInWater())));
+	return 1;
+}
```
[the remark this answers is not in the document]
Comment (bearbin) [reply src/Bindings/ManualBindings.cpp]: Yes, will do. Thanks for linking an example.

**Comment by bearbin** [comment]
> This would be better I think. That way it works properly when the pickup is between two blocks.

> I'm also wondering if IsOnFire would be enough.

I think IsOnFire would be enough, but that's uncovered another issue: SetSwimState is only called for mobs and players, so this probably breaks catching entities on fire. I'll have to call SetSwimState for all entities.

File: src/Bindings/ManualBindings.cpp [hunk 4]
```diff
@@ -3958,22 +3958,22 @@ static int tolua_cCompositeChat_UnderlineUrls(lua_State * tolua_S)
...
 {
+	// Check the params:
 	cLuaState L(tolua_S);
-
-	// Get the params:
-	cEntity * self = reinterpret_cast<cEntity *>(tolua_tousertype(tolua_S, 1, nullptr));
-	if (self == nullptr)
+	if (!L.CheckParamSelf("cEntity"))
 	{
-		LOGWARNING("%s: invalid self (%p)", __FUNCTION__, static_cast<void *>(self));
```
Comment (bearbin) [inline src/Bindings/ManualBindings.cpp]: Is removing this warning OK or do I need to replace it with a new one not depending on self?

Comment (peterbell10) [reply src/Bindings/ManualBindings.cpp]: `CheckParamSelf` give a useful error message on its own.

Comment (bearbin) [reply src/Bindings/ManualBindings.cpp]: Great. Everything else is fixed per your comments then. The build failed on AppVeyor due to a merge conflict, and travis didn't get triggered for some reason but it builds on my machine.

Comment (peterbell10) [reply src/Bindings/ManualBindings.cpp]: Can you fix the merge conflict and hopefully travis will work next time.

Comment (bearbin) [reply src/Bindings/ManualBindings.cpp]: Done.

**Comment by bearbin** [comment]
The builds of the latest commit passed, what CI are you referring to?

**Comment by peterbell10** [comment]
I don't mean this PR specifically, I mean the issues with travis builds in general.  Until #4140 gets merged, I can't merge anything.

**Pull request merged**